src/H5Oint.c

```diff
@@ -1056,14 +1055,20 @@ H5O_protect(const H5O_loc_t *loc, unsigned prot_flags, bool pin_all_chunks)
 
             /* Sanity check */
             assert(chk_proxy->oh == oh);
-            assert(chk_proxy->chunkno == chkcnt);
-            assert(oh->nchunks == (chkcnt + 1));
+
+            chunkno = chk_proxy->chunkno;
 
             /* Release the chunk from the cache */
             if (H5AC_unprotect(loc->file, H5AC_OHDR_CHK, cont_msg_info.msgs[curr_msg].addr, chk_proxy,
                                H5AC__NO_FLAGS_SET) < 0)
                 HGOTO_ERROR(H5E_OHDR, H5E_CANTUNPROTECT, NULL, "unable to release object header chunk");
 
+            if (chunkno != chkcnt)
```

Comment:
Moved these checks below the call to `H5AC_unprotect` to make sure that the chunk gets released from the cache before throwing an error